### PR TITLE
refactor: add auto reconnect to curp client

### PR DESCRIPTION
**Pull request by bsbds**
This PR includes two major changes:

* Implements an auto reconnect layer.
* Lazily build connections for client.

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

**Review by Phoenix500526 — COMMENTED** [review]
What is the principle behind changing an async method to a sync one, or vice versa? Please squash these commits into one

File: crates/curp/src/rpc/connect.rs [hunk 1]
```diff
@@ -69,85 +70,79 @@ impl FromTonicChannel for InnerProtocolClient<Channel> {
...
-            connect_to::<Client>(id, addrs, tls_config.cloned())
-                .await
-                .map(|conn| (id, conn))
-        })
-        .collect();
-    futures::StreamExt::collect::<Vec<_>>(conns_to)
-        .await
-        .into_iter()
-        .collect::<Result<Vec<_>, _>>()
+/// Creates a new connection with auto reconnect
```
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/curp/src/rpc/connect.rs:98]: Please refer to: https://github.com/hyperium/tonic/issues/1254.

File: crates/curp/src/rpc/connect.rs [hunk 2]
```diff
@@ -69,85 +70,79 @@ impl FromTonicChannel for InnerProtocolClient<Channel> {
     }
 }
 
-/// Connect to a server
-async fn connect_to<Client: FromTonicChannel>(
+/// Creates a new connection
+fn connect_to<Client: FromTonicChannel>(
```
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/curp/src/rpc/connect.rs:74]: The `change_tx.send` is in a synchronous context, where a `try_send` should never blocks. Therefore it is unecessary to use async here.

**Review by Phoenix500526 — COMMENTED** [review]
Plz squash these commits.

**Review by rogercloud — APPROVED** [review]

**Review by rogercloud — APPROVED** [review]

**Pull request merged**